### DISCUSSION TITLE
Uncaught TypeError: Cannot read properties of null (reading 'useContext") in Module Federation v^14 and v^13

### DISCUSSION THREAD

**Discussion by selvaganapathi5995** [vercel/next.js#63151] [Help]
### Link to the code that reproduces this issue

https://github.com/rakeshongithub/next13-mf-example/branches

### To Reproduce

This repository presents a practical example of how a micro application can be integrated within another Next.js 13 application. It showcases two variations of Next.js applications: one utilizing page-based routing and the other employing app-based routing.

The primary objective of this demonstration is to highlight the potential issues encountered with app-based routing and module federation.
Install dependencies

npm run install
App 1. Remote App - Custom React App

This is a micro application designed to be utilized by the host application.

To execute this application, use the following command: npm run dev --workspace @next13-mf-example/custom-react-app
App 2. Host App - Next 13 with App Dir

This is the host application that uses Next.js 13 and app-based routing.

To execute this application, use the following command: npm run dev --workspace @next13-mf-example/next13-with-app-dir
App 3. Host App - Next 13 with Page Dir

This is the host application that uses Next.js 13 and page-based routing.

To execute this application, use the following command: npm run dev --workspace @next13-mf-example/next13-with-pages-dir

### In custom-react-app/app.js-> useContext commented then app is working in both app and page router
<img width="1721" alt="Screenshot 2024-01-08 at 5 28 47 PM" src="https://github.com/rakeshongithub/next13-mf-example/assets/61230115/dafcc35c-9af4-40e7-be5a-7ae8878fe3d0">

### In custom-react-app/app.js-> useContext uncommented then app is not working in both app and page router throwing following error
<img width="1718" alt="Screenshot 2024-01-08 at 5 29 40 PM" src="https://github.com/rakeshongithub/next13-mf-example/assets/61230115/811a21d9-9331-4e8f-a884-27aa70a3eba4">

<img width="1721" alt="Screenshot 2024-01-08 at 5 31 52 PM" src="https://github.com/rakeshongithub/next13-mf-example/assets/61230115/868d4c45-1739-4bb9-92b1-f05d2286c4a0">

### Expected: 
While importing exposed reaact component that have useContext should work fine
<img width="1708" alt="Screenshot 2024-01-08 at 6 05 46 PM" src="https://github.com/vercel/next.js/assets/61230115/62235048-b4c1-4c36-b5d9-5d3c2152890a">

### Actual:
1. ### In custom-react-app/app.js-> useContext commented then app is working in page router  
Using const NextFederationPlugin = require('@module-federation/nextjs-mf') in config for module federation
<img width="1721" alt="Screenshot 2024-01-08 at 5 28 47 PM" src="https://github.com/rakeshongithub/next13-mf-example/assets/61230115/3e1a6598-fec5-45f5-8a0d-c3111ad8086a">

2. ### In custom-react-app/app.js-> useContext commented then app is not working in app router  
Using const NextFederationPlugin = require('@module-federation/nextjs-mf') in config for module federation
<img width="1666" alt="Screenshot 2024-01-08 at 6 05 54 PM" src="https://github.com/rakeshongithub/next13-mf-example/assets/61230115/38d151c8-8e44-4871-97fb-7e3a7b5c3e6f">

3. ### In custom-react-app/app.js-> useContext uncommented then app is not working in both page and app router  
Using const NextFederationPlugin = require('@module-federation/nextjs-mf') in config for module federation
<img width="1718" alt="Screenshot 2024-01-08 at 5 29 40 PM" src="https://github.com/rakeshongithub/next13-mf-example/assets/61230115/f5faf7df-3e12-40e1-8008-81e64442ed41">
<img width="1666" alt="Screenshot 2024-01-08 at 6 05 54 PM" src="https://github.com/rakeshongithub/next13-mf-example/assets/61230115/b7f64da7-c450-4093-8bdd-4af1bd80d964">

### Expected: 
While importing exposed reaact component that have useContext should work fine
<img width="1708" alt="Screenshot 2024-01-08 at 6 05 46 PM" src="https://github.com/vercel/next.js/assets/61230115/62235048-b4c1-4c36-b5d9-5d3c2152890a">

### In custom-react-app/app.js-> useContext commented then app is working in both app and page router
<img width="1721" alt="Screenshot 2024-01-08 at 5 28 47 PM" src="https://github.com/rakeshongithub/next13-mf-example/assets/61230115/dafcc35c-9af4-40e7-be5a-7ae8878fe3d0">

### In custom-react-app/app.js-> useContext uncommented then app is not working in both app and page router throwing following error
<img width="1718" alt="Screenshot 2024-01-08 at 5 29 40 PM" src="https://github.com/rakeshongithub/next13-mf-example/assets/61230115/811a21d9-9331-4e8f-a884-27aa70a3eba4">

<img width="1721" alt="Screenshot 2024-01-08 at 5 31 52 PM" src="https://github.com/rakeshongithub/next13-mf-example/assets/61230115/868d4c45-1739-4bb9-92b1-f05d2286c4a0">

### Expected: 
While importing exposed reaact component that have useContext should work fine
<img width="1708" alt="Screenshot 2024-01-08 at 6 05 46 PM" src="https://github.com/vercel/next.js/assets/61230115/62235048-b4c1-4c36-b5d9-5d3c2152890a">

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 22.6.0
Binaries:
  Node: 18.18.2
  npm: 9.8.1
  pnpm: N/A
Relevant Packages:
  next: 13.5.6, 14.0.6
  eslint-config-next: 13.5.2
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.2.2
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Not sure, App Router

### Which stage(s) are affected? (Select all that apply)

Other (Deployed)

### Additional context

There are three branches I created below mentioned are the links and each branch have different next.config.js setup.

1. Branch name: next13.5.6-webpack-mf-plugin
    Next version: 13.5.6
    Federation Module: Webpack Module Federation plugin
    Link:  https://github.com/rakeshongithub/next13-mf-example/tree/next13.5.6-webpack-mf-plugin

2. Branch name: next13.5.6-module-federation/nextjs-mf-plugin
    Next version: 13.5.6
    Federation Module: module-federation/nextjs-mf-plugin
    Link:  https://github.com/rakeshongithub/next13-mf-example/tree/next13.5.6-module-federation/nextjs-mf-plugin

3. Branch name: next14.0.3-webpack-mf-plugin
    Next version: 14.0.3
    Federation Module: Webpack Module Federation plugin
    Link:  https://github.com/rakeshongithub/next13-mf-example/tree/next14.0.3-webpack-mf-plugin

**Comment by sublayerio** [comment]
I can verify this. It happens all of a sudden.
![Screenshot 2024-01-15 at 12 56 04](https://github.com/vercel/next.js/assets/44947294/6ff3c194-ab11-458d-b86c-90106dff8ebf)

**Comment by sublayerio** [comment]
I isolated the usePathname function to see if it would make a difference, used that instead importing from 'next/navigation'.

I'm still getting the same error. It seems to me that the error occurs within usePathname called from within Next itself somewhere.

![Screenshot 2024-01-15 at 13 02 54](https://github.com/vercel/next.js/assets/44947294/42659ea6-978f-40e1-b37d-947fe53f216d)

**Comment by kelvinlouis** [comment]
We were able to avoid this issue by setting `swcMinify: false` in `next.config.js`. This is not going to be possible in v15, because turning it off is deprecated.

**Comment by Gawdfrey** [comment]
I am getting this error in a standard next project, version 14.1.0. `next.config.js` is empty.
Are the rest of you using module federation.
@kelvinlouis will attempt with `swcMinify: false`, thanks!

**Comment by Brettkgamble** [comment]
This broke for me in 14.1.1-canary.38.

Just adding `swcMinify: true` solved the issue for me.  I did not try false.

**Comment by selvaganapathi5995** [comment]
I tried with swcMinify:true but it is not working in my case.

**Comment by samcx** [comment]
@selvaganapathi5995 Hi there!

Next.js does not support Module Federation out of the box, so this is technically not an issue with core Next.js (this is requiring a custom Webpack solution) and package. Because of this, I will be moving this issue over to our <picture data-single-emoji=":nextjs:" title=":nextjs:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/nextjs/2a85d633ae594556.png" alt=":nextjs:" align="absmiddle"></picture> [discussions](https://github.com/vercel/next.js/discussions).

**Comment by stabildev** [comment]
My app with turbopack crashes continually with this error. `swcMinify:true` fixes it.
